unit/java_bytecode/java_bytecode_specialize_generics/java_bytecode_specialize_generics.cpp

```diff
@@ -0,0 +1,432 @@
+              "<java::java.lang.Integer>"));
+        const typet &new_symbol_type=specialized_method_symbol.type;
+        REQUIRE(new_symbol_type.id()==ID_code);
+
+        const code_typet &new_code=to_code_type(new_symbol_type);
+        const typet &new_method_return_type=new_code.return_type();
+        REQUIRE(!is_java_generic_parameter(new_method_return_type));
+
+        REQUIRE(new_method_return_type.subtype()==
+          symbol_typet("java::java.lang.Integer"));
+      }
+
+
+
+      THEN("Specialize a method with a generic parameter types")
+      {
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$bound_element.g:(Ljava/lang/Number;)V"));
+
+        const symbolt &generic_method_symbol=
+          *new_symbol_table.lookup(
+            "java::generics$bound_element.g:(Ljava/lang/Number;)V");
+
+        const typet &symbol_type=generic_method_symbol.type;
+        REQUIRE(symbol_type.id()==ID_code);
+
+        const code_typet &code=to_code_type(symbol_type);
+        REQUIRE(code.parameters().size()==2);
+
+        const code_typet::parametert &param=code.parameters()[1];
+        REQUIRE(is_java_generic_parameter(param.type()));
+
+        const java_generic_parametert &generic_param=
+          to_java_generic_parameter(param.type());
+        const symbol_typet &type_var=generic_param.type_variable();
+        REQUIRE(type_var.get_identifier()==
+                "java::generics$bound_element::NUM");
+
+        const symbolt &specialized_method_symbol=instantiate_generic_method(
+          generic_method_symbol,
+          {
+            { symbol_typet("java::generics$bound_element::NUM"),
+              symbol_typet("java::java.lang.Integer")
+            }
+          },
+          new_symbol_table);
+
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$bound_element.g:(Ljava/lang/Number;)V"
+              "<java::java.lang.Integer>"));
+        const typet &new_symbol_type=specialized_method_symbol.type;
+        REQUIRE(new_symbol_type.id()==ID_code);
+
+        const code_typet &new_code=to_code_type(new_symbol_type);
+        REQUIRE(new_code.parameters().size()==2);
+
+        const code_typet::parametert &new_param=new_code.parameters()[1];
+        REQUIRE(!is_java_generic_parameter(new_param.type()));
+        REQUIRE(new_param.type().subtype()==
+          symbol_typet("java::java.lang.Integer"));
+      }
+
+
+
+
+      THEN("Specialise a method with multiple generic parameters")
+      {
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$double_element.insert:"
+              "(Ljava/lang/Object;Ljava/lang/Object;)V"));
+
+        const symbolt &generic_method_symbol=
+          *new_symbol_table.lookup(
+            "java::generics$double_element.insert:"
+              "(Ljava/lang/Object;Ljava/lang/Object;)V");
+
+        const typet &symbol_type=generic_method_symbol.type;
```

Comment:
Nit: a pattern I think might be good for us to use in our unit tests is something like this: https://github.com/diffblue/cbmc/blob/develop/unit/testing-utils/require_expr.cpp So for here you could do something like:

```C++
code_typet &require_code_type(const typet &type, optionalt<size_t> num_params, optional_t<typet &> return_type)
{
  REQUIRE(type.id()==ID_code);
  const code_typet &code_type=to_code_type(type);
  if(num_params)
    REQUIRE(code_type.parameters().size()==num_params.value());
  // maybe check return type?
  return code_type;
}
```

If you think it is a good idea you might like to use it here (and put it in testing-utils/require_type)

Reply:
That does look a neat way to cleanup the tests a bit - I'll have a play with that.